File: app/src/main/java/org/michaelbel/tjgram/ui/post/adapter/GalleryAdapter.java
package org.michaelbel.tjgram.ui.post.adapter;

import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import org.jetbrains.annotations.NotNull;
import org.michaelbel.tjgram.R;
import org.michaelbel.tjgram.utils.ViewUtil;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.recyclerview.widget.RecyclerView;

public class GalleryAdapter extends RecyclerView.Adapter<GalleryAdapter.ViewHolder> {

    private static final int VIEW_TYPE_MEDIA = 1;
    private static final int VIEW_TYPE_CAMERA = 2;
    private static final int VIEW_TYPE_GALLERY = 3;

    public interface PhotoClickListener {
        void onPhotoClick(File photo);
        void onCameraClick();
        void onGalleryClick();
    }

    private List<File> photos = new ArrayList<>();
    private PhotoClickListener photoClickListener;

    public void addListener(PhotoClickListener listener) {
        photoClickListener = listener;
    }

    public void swapData(List<File> list) {
        photos.addAll(list);
        photos.add(0, new File(""));
        photos.add(photos.size() - 1, new File(""));
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return photos != null ? photos.size() : 0;
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0) {
            return VIEW_TYPE_CAMERA;
        } else if (position == photos.size() - 1) {
            return VIEW_TYPE_GALLERY;
        } else {
          return VIEW_TYPE_MEDIA;
        }
    }

    @NotNull
    @Override
    public GalleryAdapter.ViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_MEDIA) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_media, parent, false);
            return new MediaViewHolder(view);
        } else {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_camera, parent, false);
            return new BucketViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull GalleryAdapter.ViewHolder holder, int position) {
        Uri data = Uri.fromFile(photos.get(position));

        if (getItemViewType(position) == VIEW_TYPE_CAMERA) {
            BucketViewHolder viewHolder = (BucketViewHolder) holder;
            viewHolder.mTextView.setText(R.string.item_camera);
            viewHolder.mImageView.setImageDrawable(ViewUtil.INSTANCE.getIcon(viewHolder.mImageView.getContext(), R.drawable.ic_camera, R.color.icon_active));
        } else if (getItemViewType(position) == VIEW_TYPE_GALLERY) {
            BucketViewHolder viewHolder = (BucketViewHolder) holder;
            viewHolder.mTextView.setText(R.string.item_gallery);
            viewHolder.mImageView.setImageDrawable(ViewUtil.INSTANCE.getIcon(viewHolder.mImageView.getContext(), R.drawable.ic_image_outline, R.color.icon_active));
        } else {
            MediaViewHolder viewHolder = (MediaViewHolder) holder;

            Picasso.get().load(data).resize(320, 320).centerCrop()
                .placeholder(R.drawable.placeholder_rectangle).error(R.drawable.error_rectangle).into(viewHolder.imageView);
        }
    }

    abstract class ViewHolder extends RecyclerView.ViewHolder {

        final ImageView imageView;

        private ViewHolder(View itemView) {
            super(itemView);
            imageView = itemView.findViewById(R.id.imageView);
        }
    }

    private class BucketViewHolder extends ViewHolder implements View.OnClickListener {

        private final TextView mTextView;
        private final AppCompatImageView mImageView;

        private BucketViewHolder(View itemView) {
            super(itemView);
            mTextView = itemView.findViewById(R.id.text);
            mImageView = itemView.findViewById(R.id.imageView);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            int position = getAdapterPosition();
            if (position != RecyclerView.NO_POSITION) {
                if (photoClickListener != null) {
                    if (position == 0) {
                        photoClickListener.onCameraClick();
                    } else {
                        photoClickListener.onGalleryClick();
                    }
                }
            }
        }
    }

    private class MediaViewHolder extends ViewHolder implements View.OnClickListener {

        private MediaViewHolder(View itemView) {
            super(itemView);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            int position = getAdapterPosition();
            if (position != RecyclerView.NO_POSITION) {
                if (photoClickListener != null) {
                    photoClickListener.onPhotoClick(photos.get(position));
                }
            }
        }
    }
}

File: app/src/main/java/org/michaelbel/tjgram/modules/addpost/adapter/GalleryAdapter.kt
package org.michaelbel.tjgram.modules.addpost.adapter

import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.item_camera.*
import org.michaelbel.tjgram.R
import org.michaelbel.tjgram.utils.ViewUtil
import java.io.File
import java.util.*

class GalleryAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val VIEW_TYPE_MEDIA = 1
        private const val VIEW_TYPE_CAMERA = 2
        private const val VIEW_TYPE_GALLERY = 3

        private const val PHOTO_SIZE = 320
    }

    interface PhotoClickListener {
        fun onPhotoClick(photo: File)
        fun onCameraClick()
        fun onGalleryClick()
    }

    private val photos = ArrayList<File>()
    private var photoClickListener: PhotoClickListener? = null

    fun addListener(listener: PhotoClickListener) {
        photoClickListener = listener
    }

    fun swapData(list: List<File>) {
        photos.addAll(list)
        photos.add(0, File(""))
        photos.add(photos.size - 1, File(""))
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int = photos.size

    override fun getItemViewType(position: Int): Int {
        return when (position) {
            0 -> VIEW_TYPE_CAMERA
            photos.size - 1 -> VIEW_TYPE_GALLERY
            else -> VIEW_TYPE_MEDIA
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == VIEW_TYPE_MEDIA) {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.item_media, parent, false)
            MediaViewHolder(view)
        } else {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.item_camera, parent, false)
            BucketViewHolder(view)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val data = Uri.fromFile(photos[position])

        when {
            getItemViewType(position) == VIEW_TYPE_CAMERA -> {
                val viewHolder = holder as BucketViewHolder
                viewHolder.bind(R.string.item_camera, R.drawable.ic_camera)
            }
            getItemViewType(position) == VIEW_TYPE_GALLERY -> {
                val viewHolder = holder as BucketViewHolder
                viewHolder.bind(R.string.item_gallery, R.drawable.ic_image_outline)
            }
            else -> {
                val viewHolder = holder as MediaViewHolder
                viewHolder.bind(data)
            }
        }
    }

    private inner class BucketViewHolder (override val containerView: View) : RecyclerView.ViewHolder(containerView), View.OnClickListener, LayoutContainer {

        fun bind(@StringRes textId: Int, @DrawableRes iconId: Int) {
            textView.text = containerView.context.getString(textId)
            imageView.setImageDrawable(ViewUtil.getIcon(containerView.context, iconId, R.color.icon_active))
        }

        override fun onClick(v: View) {
            if (adapterPosition != RecyclerView.NO_POSITION) {
                if (adapterPosition == 0) {
                    photoClickListener!!.onCameraClick()
                } else {
                    photoClickListener!!.onGalleryClick()
                }
            }
        }
    }

    private inner class MediaViewHolder (override val containerView: View) : RecyclerView.ViewHolder(containerView), View.OnClickListener, LayoutContainer {

        fun bind(image: Uri) {
            containerView.setOnClickListener(this)
            Picasso.get().load(image).resize(PHOTO_SIZE, PHOTO_SIZE).centerCrop()
                .placeholder(R.drawable.placeholder_rectangle).error(R.drawable.error_rectangle).into(imageView)
        }

        override fun onClick(v: View) {
            if (adapterPosition != RecyclerView.NO_POSITION) {
                photoClickListener!!.onPhotoClick(photos[adapterPosition])
            }
        }
    }
}